### ISSUE TITLE
boxplot function?

### ISSUE THREAD

**Issue by danielchen26** [altair-viz/altair#664]
When will altair support boxplot?

**Comment by jakevdp** [comment]
Altair will support box plots whenever vega-lite supports boxplots.

You can follow along using the milestone for composite marks in vega-lite here: https://github.com/vega/vega-lite/milestone/56

**Comment by ajasja** [comment]
I guess vega-lite now suppoerts boxplots https://vega.github.io/vega-lite/examples/box-plot_minmax_2D_vertical_normalized.html? All the boxplot milestones are closed in https://github.com/vega/vega-lite/milestone/56.

Also there is a boxplot example in Altiar https://altair-viz.github.io/gallery/boxplot_max_min.html, so I guess this can be closed now:)

**Comment by jakevdp** [comment]
The compound marks in vega-lite haven't made it into a release yet, and the Altair example is more of a hack based on manually layering marks. I don't think that's a very viable solution, to be honest.

**Comment by justinbois** [comment]
Vega-Lite now supports boxplots by specifying the mark type as `boxplot`. I'm not familiar enough with the Altair codebase to submit a PR, but it seems like it should not be a big challenge to include it (famous last words of any developer, I know, especially when said by the guy who is not writing the code).

[Here](https://vega.github.io/editor/#/url/vega-lite/N4KABGBEAkDODGALApgWwIaQFxUQFzwAdYsB6UgN2QHN0A6agSz0QFcAjOxge1IRQyUa6ALQAbZskoAmOgDY6ABjoArWNwB2kADTgo8TQDNG1bGFAQIkCo2QB3MxctWUJ-GYDMixbudW7jAAmLGYALN56EAC+elG+UIHoeJg4TlYa6KjIZpCJySIArKGBHoEAHACM6ACc6AUA7Ipy0nKGyIbsBdIVBdWKyPByQ9mx8blJ6LDIeLCOkQkThcWllTV1jc2t7Z3dvf2Dw2YA2vMQaX7W6GKs2Tgiyn2K0vWhBQVyZfUVjWUF8X5QCjoABOjHQ7DEtwWyTAikgpzAcQR52cl2uUOU0lC2K8ZVC3V+b0q-wuQNB4MhOTy6Fh8IBSIBKMsaJuZmUDWqlW8inqzTe1TkoRJqLJYIhUPGMLhCIZfiZViB6LM9zo9RevQqHjkig80jK1WkWOFzNFFIl1NpMuN8sBV1ZOAeWPq0m5PWevWkxoVILFlJwkpp0vp1oR3qVd2UPIKeI8PTkBVjigNFU9odt5PFVImluDyLTLIxdAKFXCZX10i89WqVfeXvTvvN2aDfllzhtBeVylC9Q8Gx6FU5TwNqYB3ozfuhgbpLZDo9t4dhSlCHmqoWHA+qfWXI9Hl3Hjal0+crcs7cV9rAKvxWJ1zv1BSHFbre4bWcPVrzc47dwqdFCFW+Q0yiGADpGjYtn1NTN-QtZtj1nXdzyhX8H1CQUXTA6sexTDxIJ9M03ynD9GXzJCzF-HpYzeHtDUUUJgIAvD90InMZ0-RC7ShFUKiTeNux7D0vD+Uj8OgydWPg9iAW-S9lE1d5qh4spDTkb5-wqJjXxgpsj0sE8zhEhcVS6BMdTxFT-zQndpKgicAwkvSEJszi2XkMpFATd4y0FeMWk0gjtPfXMSK-MiHSLbEii+NceJUr5-LE+y4McqTSRcn86ArYsqL6XU1TkHsErs2DdOiJy0oXX8hjLSoB3-RppGqXCROYwKiOCuVDIvZQtQaXkPF6ZoqwrPyWq08TkrK1KRXSy8KLkAboyeHyBv1IqD3atiQo4hdI2qaNdR5Td6kJeiNLGgKJtKxFypmozlF+HltQTBqukUCo5HWljJpu6aTVm9lNxTJTwhrNdqi+tqHKm7bnMqzKBt1BaS2U4D3Uhq7iM60LZpVT5sUUzUhkUvVzpx1rMY6tsusLA08U+LVV3eAkIYuxKSqx6mcfuhG3iGD4e3xFcTIxpLrv08waeVX80LO8JNy1PVOSFNnip0znTyliM6EqdTsSG9ymvqYTyfGsWNYM7mLxEWRMN6Y6PoFOiVdNy7zapzWreQosToqYD6LLeitTLUWOY9y2dutiidQ+eMyg8X48Se0P1fDyWvfIugPuzj6XQfTlVJToKtuxyOuJ6753ITjUUwK2N6iLzbJNhiruqUJS9mLAXozj5rXfZ1OS65svpaUWNl0+ArOTAl2OIp92h89kfwoaZ5qtU3kBw8bfG+h36W7utvvE+Lomo3bxGnt3efols9Zoo-aii8LEsLkLDr-F27-p5hPS2O8CNgh1VhtPet8taLg8PifaKY+jG2GmBD+Ft07L0XPlZSvY-YLVjPGT6wDvqfz+mGC8MsBSll4l0CsHxEFpzvjzd6j19Q8iaIbN+QD+5q2Ls3UucNiFjwrFhHUDQQKKDKNQxeEceHl0yoKfanIBxNEaGuPuc8zZh3EcgyRnYIonRXF8Y2IjFIDTEVw4emiMrdFXLVVcq5XiClEXgqGN8v5EO9j2Z0ikDTGzAjxBOu8KgEIPt-XhTQhjhBUmBNGC0-EBO4a3b2eJeiVAWk1fE+pfjRKQbQtuRRwgKN2PbeiGSaHgJEL+WMjUwIUJ4tA6yaV54Wn8Zk8BFFGiQMGk1Wq2JakzXqdmRpxSM7azXu9TceoXRvxEWhIp6islQlkD2YcLxujdkavraZJil5mLAFVB8gp8QnW8ZsWezlekwn6TM8BshVJFDAs8eBK4okOPEucjZEi4mjxApuDyqkhKWIbk8+yLyUqBJcZnUGfQuhqU5J8ex7CQFAphrEw+3t+Elg9Gqfa7lYUqLdg0mJpj3k4FkI-QUmpDTdiah8E2OKB5nPxZswlqC-ZgUFn7GMPRjl1NUX0+lbzkWZxLFqJM8cEx9AHLydZwKkVBMLCmPY2JnTumrAtTlPTuV0qaYM7ZdBRmTzVNhZh1KTnqppAi-e0rQU4CqrybwD4eTKW5IUgFeLNUoOUM6JoTxug9GLBy3BcKWJmrAVq91A0PiePjoKBOWpJWIoJfyq1dBt5eUnjg2KRpnU8tdVs2QAFE5hpwYTPEsbzXxpla5A6+IhiNUAhWXUJbg0oNkF0guSNvmCv+QGqGQbnHzjbh9aMw1NQ6noryAUDbe0yUxEmPUvxt7fDeGuR5Xbnm8o0Yy222IBz3mTI1cdmaNUDJQRRYCoqDSKXomdI1XLcVZqPVsiixZtwAIKlvVV-1TmmrXbMrRPjtTI3kgTfdK7AXfvAV2ERwseh6JLB9f1NKOFfuzYymWLplwXoHBWFG2LjW3sPRcrVxKTrVQeftQ0CYJ2EL7YWZZBUWi5LMo6yjILqNmFtnRAC+J85oWrMo3DtKkP3sZQ8SBHkeK3gKoLHDN6BPbLAyGpNd5lwvHPfGGdzGLWse1viEJOEEz42ZRpstlrFyDSaOM8j7l+JGYZQm0zy4AJ+x1IOVe0m1V4cEwR49Wd4wDjtZyAajURE2b5eWolf4fmcYWj2SoBoQvrrs7+MsFiYtif5jGg9nnXkJbC6Z7yoSExweynxmTiG5PIbs4DP2xZqy-FUt8IY8Wf3hcwjhE6xZ3T0Wve52TPaqNTqTUUaidEwLfN6LWTL5WhN2avCl3Ubxj49jft1j9Jqptea2R4JQZYxVJkUuWbsTXLlZ2+OUhbdFtTvRKz1sr0h5NNqzu8eMq4qxNVQquXed2Ku5YKDqwkIzPK8mjG51bHmwBfem7l5QyN3qQcMS8NcIOxxrYhxt4TSbYyrVi8jWBHlPv3a2SqKspZvUvDHaua7oPZOo+y817VVZorVReJJ5clPkdg5p1K4zWnFx1V6GhJ486fVqnx99kzgMWgeMNMy5lbP6wc4J4yq8rxLE-PjpAwVcuXwK7FzzpLLxBXgU3M0CFovIcmebb5tJ4zWHFsm5zuNtncu-nei0aM-5mhYipazEDFoHelqdyZ0Iqo1SVk9V8Hi+Izdo7swoFoMXuxxy928aPtPmk6pVZQoCG8HxI-l9TxXiWfMG+BnRNU8cXip654HnnDw37A2x8b74PuEMgP942zbf5YH7S+N4V4a4dRV8d6FkzLSijeHjO8QRzQte2Tb4XqHOskx+2Nk1DCan33s4L7rmSZTlq-21Jx8IKYh8B5HzzrbsjqyCpLEtta9uF+j51UtqswilnKXg-x27j+9dZx1PQlcfGF4OiU-DvFDVUAUIA8sOuRqUAydMKOaLOaNSyJ7NSbeTtVvFidveA++TKKMAqKzL4BoNcTffPb-HfBAlUPbGRHJLwLwbdDAr-efCg3GaHPvGxB8BMD4THOA-rBAy-boXyJWQCLcXgljAbCyEmNTPzfUD6MQzTGSLbZoIcTUBOWqIReQ7nAbSNFJOOCPFoT-UrZg83C-aRDBbwLeKeLjTQmvGSBQTUdSTjeubUUg7XbfEwmSK5RqBoHUOVZfIxB-Fg3aDPP2eiJoV4MNPqGw8-GSUpJNDyTcL5NcPUHRaInLJ-EsJhBwqiMsKsNIunWQVGDyGKBSAaRSFbLfcgjwygtg3I0PFoOVLrfI8DJNNJK7WiZbcyZowjVUApTkDkZcZfbo7zeSXUYCaxILAI33bMbAvggGJfXsclZIkRcIwwm7YwmPXLWQERD4bw8jILN3YYrZYPD4ZfP2ZMdXe-aYmEWY8Q-gpQczfnSBZSNCYnI4xlLbE6JMCwxPACMTPPNwqozYoPeI-abeB1WHRVCosgkBDwH-Own2BaPUFXXIuxXeOEoItuNGNCSBAmC+MJdE+EhA-XIYaMACDeY2HeSbDE6o2aYPTxZ0BSMk4sHkQkzElFDUIOPEWDShWfUSMrGk4EnnX7Iock7BE+eqVwufFiQUtPLVE4gUGfOiBoMNY2aEwE2Eok+Y+qDyMsYsT4BZJ7Nk2khcLbNCPNAaFMBqDlY0oUmSYPBdSBPiLBZ4E-akrUhcQomMZZMlLEQ0DLa4mkWU6vGIhAh6J4J6AUUmfWTcW0uUlBPGfUZ7TCeiE3YCOMkM9I4UnVfUuiTHZ0UddU6UqGYM4fLM+09uLoE6L4QUOHAE4s8SUss-cshA22A0ERJYnOKhd09k1yTUOdAmV4TedAtYqnAUj0i8aoTKTkHvbeRSCw0cyozU3s8LXsA6SfdyGOAMzAksicqEMoeQSAk+JGNUxiHsk0i8XNfE96f8aFJMKkwMsAJssAovBOXw8GZ4czFJDMssunE414KueqFMKBRcmEmUvctjCAlcasJofU8fCbR858nA4IjyfnJ2V4Z4G8Svc8u0yg22IoEVZWSuTjH85sgopQTccbZTACSFIs-k5ci8qEM0v2WMZVZSboC7Uil83LOItCBofUCeUUmtPkz9J8iCnALbWHbwFobeZ+QWLi5CycrOROMdT4G8FwhSuY00+QeRYmcI6sBMKU+i8ClcxcVSEjJoWRDyHkbpMchi3C+YlcHUCsZccGbUaszSu42aSSp9YObBeuPUUCjUkyxiswOPDc2syXB5WypckKhyhcA8zDQ-csLwQREStbJCrSy8pNZvUVQ0Q0YcT4TyhQhA37XUHoJ6LwbsGKYqrQhA+k3OXkSodA-fOi0SzKryhKzKD-MlNUERU+MmHcxs8SsABQQrboL4Y6R6asWq2w+4yDTIyi7uX+Wa0M2aEU6pMyVy7UGMVals7ynVSMiM3oYAgLPaunEQLbBoFMejV4HiTBAqc69PQBPYJzLKMYp6rVOInxF0QCW85SM8xCkaqc9fVhIVRzPInC48eYAAXVGD0EgGQA0AMECEYA0FMFSHmEgAMDEG4GBDmGkmMGQDEECByGlONEgDwAAE9CAJQNBuBUA0arhdITxIAAAPAmi4ImkmsmlyCm6m2mnIAAR1YHQA0DwGYCSEYCoBZpJEgCps5tRG5tJv9HJoREppprpoZqZrEBZvhqsAwGBAAGsch2BuA2bCBca8B4QoggA) is a simple example. I generated the JSON using the following Python code.

```python
import numpy as np
import pandas as pd
import altair as alt

df = pd.DataFrame(data={'data ' + str(i): np.random.normal(*musig, size=50) 
                           for i, musig in enumerate(zip([0, 1, 2, 3], [1, 1, 2, 3]))}).melt()

chart_json = alt.Chart(df
    ).mark_point(
    ).encode(
        x='value:Q',
        y='variable:N',
        color='variable:N'
    ).to_json()

chart_json = chart_json.replace('point', 'boxplot')

with open('sample_boxplot.vl.json', 'w') as f:
    f.write(chart_json)
```

**Comment by jakevdp** [comment]
Full box plot support is part of vega-lite 3.0, which has not yet been released. We will support box plots in Altair once that release is available.

**Comment by justinbois** [comment]
Ah, didn't notice that 3.0.0 is not an official release yet. Thanks for the reply, and looking forward to boxplot incorporation in Altair!

**Comment by elmbeech** [comment]
> ```python
> import numpy as np
> import pandas as pd
> import altair as alt
> 
> df = pd.DataFrame(data={'data ' + str(i): np.random.normal(*musig, size=50) 
>                            for i, musig in enumerate(zip([0, 1, 2, 3], [1, 1, 2, 3]))}).melt()
> 
> chart_json = alt.Chart(df
>     ).mark_point(
>     ).encode(
>         x='value:Q',
>         y='variable:N',
>         color='variable:N'
>     ).to_json()
> 
> chart_json = chart_json.replace('point', 'boxplot')
> 
> with open('sample_boxplot.vl.json', 'w') as f:
>     f.write(chart_json)
> ```

Hi Justin and Jake,

This is a cool hack!
Now I wonder if there is an easy way to programmatically (by some python code)
convert this json to a png.

Reading the altair documentation I understand that there has to run a  webbserver (or chromedriver)
as backbone and that this webbrowser will use vega-embed to translate the vega json to a plot.
But I struggle to figure out how to programmatically achieve this.
Because of course obj = alt.Chart.from_json(chart_json) will in this case through a ValidationError.
Is there a workaround to achieve this? 

I am afraid I don't understand enough to help to make altair vesa3 compatible - because this is what really should be done.  But maybe you can guide me there to the right place. 
That would be ok too.

Thank you, Elmar

**Comment by jakevdp** [comment]
You'd need to open a webpage in a headless browser, load an HTML file that displays a chart using the vega-lite 3 prerelease, inject the JSON, and then tell chromedriver to click the "Save PNG" button.

Or just manually load the json in http://vega.github.io/editor (which supports vega-lite 3) and click the button.

Or wait for Altair version 3 :smile:

**Comment by elmbeech** [comment]
Thank you Jake for this prompt answer and sorry for my late reply. 

I was able to install on my operating system npm and with the npm vega and vega-lite. 
And under node_modules/vega-lite/bin I can see the vl2png command. 
But some how I don't get vl2png running.
I get the error that it not can find vg2png - even I have installed vega.

Anyhow, for the moment copy pasting the json description into the  vega editor does a ok job.
I think I just wait for altair version3. :' ).

Thank you for all your help (and your Python Data Science Handbook)!
Best, Elmar
